lib/class_relationships.rb

```diff
@@ -0,0 +1,67 @@
+# Think of these toy objects as the singletons
+ToyObject = BasicObject.new
+
+def ToyObject.to_s
+  inspect
+end
+
+def ToyObject.inspect
+  "ToyObject"
+end
+
+def ToyObject.toy_class
+  ToyClass
+end
+
+def ToyObject.toy_kind_of?(klass)
+  superclass = toy_class
+  while superclass
```

Comment:
Ah! Well, in the real Ruby world, I know that `BasicObject` is at the top of the hierarchy, so in that scenario we could stop searching up the ancestor chain once we reach `BasicObject`. In our case, we've established that `ToyObject` is at the top of the hierarchy in our toy model, so presumably once we've reached `ToyObject` we could stop.

Maybe something like this (demonstrated for `ToyObject`, but would be the same for the other two classes):
```ruby
  def ToyObject.toy_kind_of?(klass)
    superclass = toy_class
    while superclass
      return true if klass == superclass
      break if superclass == ToyObject
      superclass = superclass.toy_superclass
    end
    false
  end
```  

We could also add the `klass == superclass` check after the loop, and then check whether we're at `ToyObject` right in the while loop:
```ruby
  def ToyObject.toy_kind_of?(klass)
    superclass = toy_class
    while superclass != ToyObject
      return true if klass == superclass
      superclass = superclass.toy_superclass
    end
    klass == superclass
  end
```

Reply:
Maybe we ought to write a test that would (currently) fail, something like.
```ruby
 specify 'Object is not kind of other type' do
  AnotherType = BasicObject.new
  refute call_method(_Object, "toy_", :kind_of?, AnotherType)
end
```

Reply:
> ```ruby
> break if superclass == ToyObject
> ```

This is definitely a valid solution! The downside is that we’re hardcoding some knowledge about a specific object — “`ToyObject` is special” — into all the implementations of `#toy_kind_of?`, which works fine but might make it more difficult to understand & maintain this code in the hypothetical future.

As Sandi Metz is fond of saying: depend upon abstractions, not concretions. What behavioural aspect of `ToyObject` makes it special here, and how do we check for that aspect rather than baking in an assumption about a specific singleton?

Reply:
> ```ruby
> specify 'Object is not kind of other type' do
>   AnotherType = BasicObject.new
>   refute call_method(_Object, "toy_", :kind_of?, AnotherType)
> end
> ```

That’s a good idea — ultimately I think it’d help us to keep adding these “sense check” unit tests to make sure we’re not missing any edge cases.

(In practice the above test wouldn’t be suitable because the real `#kind_of?` can only accept a class or module as its argument, so we’d want `AnotherType` to be `Class.new` or `Module.new` rather than `BasicObject.new`, but that’s a detail.)

Reply:
> What behavioural aspect of ToyObject makes it special here

So the other thought I had when writing this was that we haven't written a `.toy_superclass` method for `ToyObject`, because it doesn't have a superclass! Maybe we could write `ToyObject.toy_superclass` to return `false`, which would instead cause `while superclass` to exit once we've reached ToyObject, because `superclass` would be falsey? I was worried there though that the details of `.toy_kind_of` were dependent on `.toy_superclass` returning a falsey value for `ToyObject`. Is this what you had in mind, or am I still missing the target on this one? 😆

Reply:
> I think it’d help us to keep adding these “sense check” unit tests

Cool, maybe we can work on this IRL to make sure it makes sense!

Reply:
> Is this what you had in mind, or am I still missing the target on this one? 😆

Yeah, that’s the direction I was thinking of! What does the real `BasicObject` do?

Reply:
Looks like `BasicObject.kind_of?` comes from `Kernel`:
```bash
irb(main):003:0> BasicObject.method(:kind_of?)
=> #<Method: #<Class:BasicObject>(Kernel)#kind_of?(_)>
```

With the implementation here:
https://github.com/ruby/ruby/blob/c0c2b31a35e19a47b499b57807bc0a0f9325f6d3/object.c#L789-L798

Which makes use of:
https://github.com/ruby/ruby/blob/c0c2b31a35e19a47b499b57807bc0a0f9325f6d3/object.c#L800-L809

Interestingly, the logic looks familiar! 😉 
```c
    while (cl) {
	if (cl == c || RCLASS_M_TBL(cl) == RCLASS_M_TBL(c))
	    return cl;
	cl = RCLASS_SUPER(cl);
    }
```

So it uses `RCLASS_SUPER` to move up the inheritance chain. I think `RCLASS_SUPER` is aliased to `rb_class_get_superclass` FWICT here:
https://github.com/ruby/ruby/blob/cb4e2cb55a59833fc4f1f6db2f3082d1ffcafc80/include/ruby/internal/core/rclass.h#L46

And [the implementation](https://github.com/ruby/ruby/blob/c0c2b31a35e19a47b499b57807bc0a0f9325f6d3/object.c#L2021-L2064) for `rb_class_get_superclass` helpfully tells us this:
```
 *  Returns nil when the given class does not have a parent class:
 *
 *     BasicObject.superclass   #=> nil
 *
 *--
```

With the implementation like so:
```c
    VALUE super = RCLASS_SUPER(klass);


    if (!super) {
	if (klass == rb_cBasicObject) return Qnil;
	rb_raise(rb_eTypeError, "uninitialized class");
    }
```

So this looks like what we've done here! 🤓

Reply:
Cool! So it sounds like `#superclass` should return `nil`?

Reply:
Yeah! I realized it should actually be `nil` over `false` a little while ago, and pushed the commit for it already :D https://github.com/adrianna-chang-shopify/ruby-object-model/pull/19/commits/a30ada5fc8b5514c56dfa54bee5d68b7cdae219e

Reply:
Nice. Way ahead of me! 😂